curp/src/server/raw_curp/log.rs

```diff
@@ -175,30 +184,54 @@ impl<C: 'static + Command> Log<C> {
             (self.base_index, self.base_term)
         } else {
             let entry = self.get(i - 1).unwrap_or_else(|| {
-                unreachable!(
-                    "system corrupted, get log[{}] when we only have {} log entries",
-                    i - 1,
-                    self.last_log_index()
-                )
+                unreachable!("get log[{}] when base_index is {}", i - 1, self.base_index)
             });
             (entry.index, entry.term)
         }
     }
 
     /// Reset log base by snapshot
     pub(super) fn reset_by_snapshot_meta(&mut self, meta: SnapshotMeta) {
-        self.base_index = meta.last_included_index.numeric_cast();
+        self.base_index = meta.last_included_index;
         self.base_term = meta.last_included_term;
-        self.last_applied = meta.last_included_index.numeric_cast();
-        self.commit_index = meta.last_included_index.numeric_cast();
+        self.last_as = meta.last_included_index;
+        self.last_exe = meta.last_included_index;
+        self.commit_index = meta.last_included_index;
         self.entries.clear();
     }
+
+    /// Restore log entries, provided entires must be in order
+    pub(super) fn restore_entries(&mut self, entries: Vec<LogEntry<C>>) {
+        self.entries = VecDeque::from(entries);
+        self.compact();
+    }
+
+    /// Compact log
+    pub(super) fn compact(&mut self) {
+        let compact_from = if self.last_as >= self.entries_cap.numeric_cast() {
+            self.last_as - self.entries_cap.numeric_cast::<u64>()
+        } else {
+            return;
+        };
+        while self
+            .entries
+            .front()
+            .map_or(false, |e| e.index <= compact_from)
+        {
+            #[allow(clippy::unwrap_used)] // checked
+            let e = self.entries.pop_front().unwrap();
+            self.base_index = e.index;
+            self.base_term = e.term;
+        }
```

Comment:
This doesn't look very efficient, I recommend use `split_off` or `drain`

Reply:
Since we call compact every time we apply, this loop will run only a few times(once for most of the times). `split_off` or `drain`, if I am not mistaken, will allocate new space elements. So I think pop_front is a more efficient approach.

Reply:
Shell we trigger compact so often?

Reply:
I think it's fine. It's just a pop_front operation in memory when a new log entry is pushed. I think only disk log entries compaction needs to be triggered asynchronously.